Feat/#874 merge schemas

ciaranschutte opened the pull request:
- adds merging remote schema functionality, uses graphql tools
- adds unit tests for merging schemas (also adds watch test command)
- merging into feature branch

ciaranschutte commented on modules/server/src/network/__tests__/index.test.ts:
```diff
@@ -0,0 +1,122 @@
+// @ts-nocheck
+// graphql types have readonly modifiers all the way down
```
disabling TS on this file. It's a test file so it won't break the build. 
It's a larger piece of work to type because all the graphql lib types have `readonly` modifiers , which means util funcs like `pick` and `omit` don't work. The full types are too broad as we're using a specific subset of the functionality.

ciaranschutte commented on modules/server/src/network/__tests__/index.test.ts:
```diff
@@ -0,0 +1,122 @@
+
+/**
+ * At no point do we modify the type definitions within the tests
+ */
```
schema outside of test fixtures as we don't modify the schema inputs

ciaranschutte commented on modules/server/src/network/index.ts:
```diff
@@ -108,14 +115,31 @@ const fetchRemoteSchemas = async ({
+	// filter remote connections with valid schemas
+	const gqlSchemas = networkConfigsWithSchemas
+		.filter((config) => config.schema !== null)
+		.map((config) => config.schema as GraphQLSchema);
```
can change this to a filter func if you'd like, I feel there's a cleaner way to do this overall.

joneubank commented on modules/server/src/network/__tests__/index.test.ts:
```diff
@@ -0,0 +1,131 @@
+
+const networkSchemas = [buildSchema(typeDefsA), buildSchema(typeDefsB)];
+
+describe('network aggregation', () => {
+	test('it should have defined GQL Object types', () => {
```
Unless this is the pattern used in tests throughout arranger, it would be nice to use proper capitalization.

ciaranschutte replied:
lower case is consistent with the project
"Object" is capitalized because it's a specific GQL type not a js object

joneubank commented on modules/server/src/network/__tests__/index.test.ts:
```diff
@@ -0,0 +1,131 @@
+	/**
+	 * example: multiple remote connections have the same field
+	 */
+	test('it should de-deupe FieldDefinition', () => {
```
de-dupe typo

joneubank commented on modules/server/src/network/__tests__/index.test.ts:
```diff
@@ -0,0 +1,131 @@
+import { buildSchema } from 'graphql';
+import { createNetworkAggregationTypeDefs } from '../schema';
+
+/**
+ *
+ * @param typeDefs - Array of DefinitionNodes representing the schema type definitions
+ * @param typeName - ObjectTypeDefinition
+ * @returns true if type is defined
+ */
+const isTypeDefined = (typeDefs, typeName) => {
+	return typeDefs.some(
+		(definition) =>
+			definition.kind === 'ObjectTypeDefinition' && definition.name.value === typeName,
+	);
+};
+
+/**
+ *
+ * @param typeDefs - Array of DefinitionNodes representing the schema type definitions
+ * @param parent - ObjectTypeDefinition name
+ * @param child - FieldDefinition name
+ * @returns true if field is defined
+ */
+const isFieldDefined = (typeDefs, parent, child) => {
+	const parentTypeDefFields = typeDefs.find(
+		(definition) => definition.kind === 'ObjectTypeDefinition' && definition.name.value === parent,
+	).fields;
+	return parentTypeDefFields.some(
+		(definitions) => (definitions.kind = 'FieldDefinition' && definitions.name.value === child),
+	);
+};
```
Can probably move some test utilities to separate files to organize this code better.

joneubank commented on modules/server/src/network/__tests__/index.test.ts:
Ideas for things that should be tested:
- merging many schemas (we only test the case with 2)
- no issues merging schemas with no overlapping types/fields
- all types/fields from each schema can be found in output schema

You already kind of test this last condition but didnt declare it explicitly, you just wrote that the test was checking if "types are defined".

ciaranschutte commented on modules/server/src/network/schema.ts:
```diff
@@ -0,0 +1,21 @@
+ * @returns
+ */
+export const createNetworkAggregationTypeDefs = (gqlTypes: GraphQLSchema[]): DocumentNode => {
+	const options: Omit<Partial<Config>, 'commentDescriptions'> = {
+		forceSchemaDefinition: true,
+	};
```
removed, I think it's better to leave it out for now.

ciaranschutte commented:
@joneubank addressed feedback. 
- added each additional test you suggested. 
- added additional test for type collisions (currently error thrown is expected, we can discuss correct behaviour but hopefully this is enough of a test to merge this PR)
- added documentation to clarify some points
- ensured naming of tests is consistent with lower case
- created utils folder, created fixtures folder
- naming the schemas for tests is difficult so I added comment blocks describing their intent and what the objects contain
